### PR TITLE
do not pull conda 4.13 from defaults

### PR DESCRIPTION
**Pull request by wolfv**
Checklist
* [ ] Added a ``news`` entry

### CODE REVIEW

**Comment by BastianZim** [comment]
Tested in https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=513968&view=logs&j=171a126d-c574-5c8c-1269-ff3b989e923d&t=b5c76c08-4f10-5244-840d-58e98b91712a
Commit: https://github.com/conda-forge/dwave-preprocessing-feedstock/pull/4/commits/5361882da95d1dfda58733378474a79269c3fee9

File: conda_smithy/templates/azure-pipelines-win.yml.tmpl [hunk 1]
```diff
@@ -56,9 +56,9 @@ jobs:
 {%- endif %}
     - task: CondaEnvironment@1
       inputs:
-        packageSpecs: 'python=3.9 conda-build conda pip {{ BOA }}{{ " ".join(remote_ci_setup) }}' # Optional
-        installOptions: "-c conda-forge"
-        updateConda: true
+        packageSpecs: 'python=3.9 conda-build conda=4.12 pip {{ BOA }}{{ " ".join(remote_ci_setup) }}' # Optional
```
Comment (isuruf) [inline conda_smithy/templates/azure-pipelines-win.yml.tmpl:59]: Why is this necessary?

Comment (BastianZim) [reply conda_smithy/templates/azure-pipelines-win.yml.tmpl:59]: Do you mean the general PR or this line? It's supposed to fix this error: https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=513944&view=logs&jobId=171a126d-c574-5c8c-1269-ff3b989e923d&j=171a126d-c574-5c8c-1269-ff3b989e923d&t=b5c76c08-4f10-5244-840d-58e98b91712a

Comment (isuruf) [reply conda_smithy/templates/azure-pipelines-win.yml.tmpl:59]: this line

Comment (BastianZim) [reply conda_smithy/templates/azure-pipelines-win.yml.tmpl:59]: https://gitter.im/conda-forge/conda-forge.github.io?at=6294e1824aa6c31dca2977b1

In the previous tests, it wasn't automatically downgraded and failed instead.

Comment (wolfv) [reply conda_smithy/templates/azure-pipelines-win.yml.tmpl:59]: I've simply pinned conda to 4.12 to update it (if necessary), but not beyond 4.12. 

The `updateConda: true` option does not take into account any additional arguments passed in the `installOptions` so we cannot restrict the action to `conda-forge` and it will update to 4.13 from defaults.

Later, conda cannot downgrade because I think the solver restricts it (but mamba & boa currently require 4.12).

Comment (isuruf) [reply conda_smithy/templates/azure-pipelines-win.yml.tmpl:59]: Please add any conda conditions to `conda-forge-ci-setup` package. We don't want to keep updating this in conda-smithy and rerendering every other feedstock

Comment (wolfv) [reply conda_smithy/templates/azure-pipelines-win.yml.tmpl:59]: this is fixing an issue that currently affects all feedstocks. 

It's not really possible otherwise to restrict conda to 4.13 since defaults is still available at this stage of the Azure Pipeline.
I think what you are suggesting is a bigger change.

Comment (wolfv) [reply conda_smithy/templates/azure-pipelines-win.yml.tmpl:59]: Or, if you prefer, I can remove the version pinning and just add the `--strict-channel-priority` and that should (hopefully) do the trick to only update to `conda 4.12` max.

Comment (wolfv) [reply conda_smithy/templates/azure-pipelines-win.yml.tmpl:59]: this is what the `updateConda: true` does: https://github.com/microsoft/azure-pipelines-tasks/blob/2135c795837e3692569b3baa4cccefc9f4c9b94d/Tasks/CondaEnvironmentV1/conda_internal.ts#L93-L102

Comment (isuruf) [reply conda_smithy/templates/azure-pipelines-win.yml.tmpl:59]: > Or, if you prefer, I can remove the version pinning and just add the --strict-channel-priority and that should (hopefully) do the trick to only update to conda 4.12 max.

Do this and also add the `conda=4.12`  to `conda-forge-ci-setup`. I don't understand why you think this wouldn't work.

File: conda_smithy/templates/azure-pipelines-win.yml.tmpl [hunk 2]
```diff
@@ -56,9 +56,9 @@ jobs:
...
+        installOptions: "-c conda-forge --strict-channel-priority"
```
Comment (hmaarrfk) [inline conda_smithy/templates/azure-pipelines-win.yml.tmpl:60]:
```suggestion
        installOptions: "-c conda-forge --override-channels"
```

Can we not drop defaults early in the setup process? I thought we are quite independent from them now.

File: conda_smithy/templates/azure-pipelines-win.yml.tmpl (hunk 1 above)
Comment (wolfv) [reply conda_smithy/templates/azure-pipelines-win.yml.tmpl:59]: @isuruf you are referring to this line in the meta.yaml, right? https://github.com/conda-forge/conda-forge-ci-setup-feedstock/blob/c7a5453483f65d0e307249250bb180e033d41d4f/recipe/meta.yaml#L62

I am sure I can add the same constraint there.

File: conda_smithy/templates/azure-pipelines-win.yml.tmpl [hunk 3]
```diff
@@ -57,8 +57,8 @@ jobs:
     - task: CondaEnvironment@1
       inputs:
         packageSpecs: 'python=3.9 conda-build conda pip {{ BOA }}{{ " ".join(remote_ci_setup) }}' # Optional
-        installOptions: "-c conda-forge"
-        updateConda: true
+        installOptions: "-c conda-forge --override-channels"
+        updateConda: false
```
Comment (hmaarrfk) [inline conda_smithy/templates/azure-pipelines-win.yml.tmpl:61]: we should be able to update conda with `--override-channels` now right?

Comment (wolfv) [reply conda_smithy/templates/azure-pipelines-win.yml.tmpl:61]: If the action would take that into account, then yes, but I do not think it does: https://github.com/microsoft/azure-pipelines-tasks/blob/2135c795837e3692569b3baa4cccefc9f4c9b94d/Tasks/CondaEnvironmentV1/conda_internal.ts#L93-L102

Note this other line: https://github.com/microsoft/azure-pipelines-tasks/blob/2135c795837e3692569b3baa4cccefc9f4c9b94d/Tasks/CondaEnvironmentV1/conda_internal.ts#L121-L123

Comment (hmaarrfk) [reply conda_smithy/templates/azure-pipelines-win.yml.tmpl:61]: ok

File: conda_smithy/templates/azure-pipelines-win.yml.tmpl (hunk 1 above)
Comment (wolfv) [reply conda_smithy/templates/azure-pipelines-win.yml.tmpl]: Btw. that constraint already exists in the `boa` package now (which is why we're seeing this error). conda is not "downgradeable" after an update.

Comment (isuruf) [reply conda_smithy/templates/azure-pipelines-win.yml.tmpl]: @wolfv, I'm only talking about this line as I mentioned previously. Not the other lines where install options and updateConda is changed.

File: conda_smithy/templates/azure-pipelines-win.yml.tmpl [hunk 4]
```diff
@@ -54,12 +54,13 @@ jobs:
...
-        packageSpecs: 'python=3.9 conda-build conda pip {{ BOA }}{{ " ".join(remote_ci_setup) }}' # Optional
-        installOptions: "-c conda-forge"
-        updateConda: true
-      displayName: Install conda-build and activate environment
+    - powershell: Write-Host "##vso[task.prependpath]$env:CONDA\Scripts"
+      displayName: Add conda to PATH
+
+    - script: |
+        call activate base
+        conda.exe install 'python=3.9' conda-build conda pip {{ BOA }}'{{ " ".join(remote_ci_setup) }}' -c conda-forge --strict-channel-priority
```
Comment (isuruf) [inline conda_smithy/templates/azure-pipelines-win.yml.tmpl:62]: @wolfv, I've moved to calling conda directly as the azure task was problematic. What do you think?

Comment (wolfv) [reply conda_smithy/templates/azure-pipelines-win.yml.tmpl:62]: Looks great!

Comment (isuruf) [reply conda_smithy/templates/azure-pipelines-win.yml.tmpl:62]: Can you add a news entry? Then we can merge.

Comment (wolfv) [reply conda_smithy/templates/azure-pipelines-win.yml.tmpl]: I added a news entry but I think your other fix looks even better. Thanks!

**Comment by isuruf** [comment]
Superceded by #1631
